File: automation-test-playwright/pages/admin/AdminPropertyRequestListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class AdminPropertyRequestListPage extends CrudListPage {
  protected readonly path = "/admin/property-request/list";
  readonly tableBody: Locator;
  private readonly table: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.tableBody = this.page.locator("#requestTableBody");
    this.table = new TableComponent(page, "#requestTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/property-request\/list/);
    await this.waitForVisible(this.tableBody);
  }

  async waitForTableData(): Promise<void> {
    await this.table.waitForDataOrEmpty();
  }

  /**
   * Applies the status filter.
   */
  async filterByStatus(status: "PENDING" | "APPROVED" | "REJECTED" | "CANCELLED" | ""): Promise<void> {
    await this.page.locator("#statusFilter").selectOption(status);
  }

  rowByRequestId(requestId: number): Locator {
    return this.firstVisible(this.page.locator("#requestTableBody tr").filter({ hasText: `#${requestId}` }));
  }

  async openDetail(requestId: number): Promise<void> {
    await this.rowByRequestId(requestId).locator(`a[href="/admin/property-request/${requestId}"]`).click();
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminPropertyRequestListPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.suntower.core.BasePage;

public class AdminPropertyRequestListPage extends BasePage {
  private final By tableBody = css("#requestTableBody");

  public AdminPropertyRequestListPage(WebDriver driver) {
    super(driver);
  }

  public void waitForLoaded() {
    waitForUrlContains("/admin/property-request/list");
    waitForVisible(tableBody);
  }

  public void filterByStatus(String status) {
    selectByValue(css("#statusFilter"), status);
    waitForTableData();
  }

  public void waitForTableData() {
    waitForCondition(() -> count(css("#requestTableBody tr")) > 0, "Property request table did not load.");
  }

  public void waitForRowVisible(Long requestId) {
    waitForCondition(() -> isVisible(rowByRequestId(requestId)), "Property request row was not visible: " + requestId);
  }

  public void openDetail(Long requestId) {
    click(By.cssSelector("a[href='/admin/property-request/" + requestId + "']"));
  }

  private By rowByRequestId(Long requestId) {
    return By.xpath("//tbody[@id='requestTableBody']//tr[.//td[contains(normalize-space(.), '#" + requestId + "')]]");
  }
}
